src/algorithms/highlevel/crosssimilaritymatrix.cpp

```diff
@@ -0,0 +1,136 @@
+  int increment = frameStackSize * frameStackStride;
+  std::vector<std::vector<Real> > stackedFrames;
+  stackedFrames.reserve(frames.size() - increment);
+  std::vector<Real> stack;
+  for (size_t i=0; i<(frames.size() - increment); i+=frameStackStride) {
+    stopIdx = i + increment;
+    for (size_t startTime=i; startTime<stopIdx; startTime+=frameStackStride) {
+      stack.insert(stack.end(), frames[startTime].begin(), frames[startTime].end());
+    }
+    stackedFrames.push_back(stack);
+    stack.clear();
+  }
+  return stackedFrames;
+}
+
+
+void CrossSimilarityMatrix::compute() {
+  // get inputs and output
+  std::vector<std::vector<Real> > queryFeature = _queryFeature.get();
+  std::vector<std::vector<Real> > referenceFeature = _referenceFeature.get();
+  std::vector<std::vector<Real> >& csm = _csm.get();
+
+  if (queryFeature.empty())
+    throw EssentiaException("CrossSimilarityMatrix: input queryFeature array is empty.");
+  if (referenceFeature.empty())
+    throw EssentiaException("CrossSimilarityMatrix: input referenceFeature array is empty.");
+
+  // construct a new vector by stacking the input features by an specified 'frameStackStride' and 'frameStackSize'
+  std::vector<std::vector<Real> >  queryFeatureStack = stackFrames(queryFeature, _frameStackSize, _frameStackStride);
+  std::vector<std::vector<Real> >  referenceFeatureStack = stackFrames(referenceFeature, _frameStackSize, _frameStackStride);
+
+  std::vector<std::vector<Real> > pdistances;
+  // pairwise euclidean distance
+  pdistances = pairwiseDistance(queryFeatureStack, referenceFeatureStack);
+
+  // check whether to binarize the euclidean cross-similarity matrix using the given threshold kappa
+  if (_toBinary == true) {
+    std::vector<std::vector<Real> > tpDistances = transpose(pdistances);
+    size_t queryRows = pdistances.size();
+    size_t referenceRows = pdistances[0].size();
+
+    std::vector<Real> thresholdX(queryRows);
+    std::vector<Real> thresholdY(referenceRows);
+
+    std::vector<std::vector<Real> > outputSimMatrix;
+    outputSimMatrix.assign(queryRows, std::vector<Real>(referenceRows));
+    // construct the binary output similarity matrix using the thresholds computed along the queryFeature axis
+    for (size_t k=0; k<queryRows; k++) {
+      thresholdX[k] = percentile(pdistances[k], _kappa*100);
+      for (size_t l=0; l<referenceRows; l++) {
+        if ((thresholdX[k] - pdistances[k][l]) > 0) {
+          outputSimMatrix[k][l] = 1;
+        }
+        else if ((thresholdX[k] - pdistances[k][l]) <= 0) {
+          outputSimMatrix[k][l] = 0.;
+        }
+      }
+
+    }
+    // update the binary output similarity matrix by multiplying with the thresholds computed along the referenceFeature axis
+    for (size_t j=0; j<referenceRows; j++) {
+      thresholdY[j] = percentile(tpDistances[j], _kappa*100);
+      for (size_t i=0; i<queryRows; i++) {
+        if ((thresholdY[j] - pdistances[i][j]) > 0) {
+          outputSimMatrix[i][j] *= 1; 
```

Comment:
Skip or set to zero instead of ```*=1``` and ```*=0```. You don't even need to check for the condition to do ```*=1```, just check for the condition to assign to zero.